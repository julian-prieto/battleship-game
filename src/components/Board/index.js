import { useState } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './styles.module.scss';

const boardAxis = {
  columns: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'],
  rows: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'],
};

export default function Board({ board, userBoard, onSinkTry, onSaveGame: handleSaveGame }) {
  const size = board.length;
  const [showUserBoard, setShowUserBoard] = useState(true);

  return (
    <div className={styles.main}>
      <div data-testid="board" className={styles.board}>
        <div className={styles.head}>
          <div className={styles.axis} />
          {boardAxis.columns.map((value) => (
            <div key={value} className={styles.axis}>
              {value}
            </div>
          ))}
        </div>
        {Array.from({ length: size }).map((_, y) => (
          <div key={`row-user-${y}`} className={styles.row}>
            <div key={`col-rows-${y}`} className={styles.axis}>
              {boardAxis.rows[y]}
            </div>
            {showUserBoard
              && Array.from({ length: size }).map((_, x) => (
                <div
                  key={`col-user-${x}-${y}`}
                  data-testid={`col-user-${x}-${y}`}
                  className={classnames(styles.column, {
                    [styles.failed]: userBoard[y][x][0] === 'X',
                    [styles['sunk-1']]: userBoard[y][x] === 'O1',
                    [styles['sunk-2']]: userBoard[y][x] === 'O2',
                    [styles['sunk-3']]: userBoard[y][x] === 'O3',
                    [styles['sunk-4']]: userBoard[y][x] === 'O4',
                    [styles.partial]: Boolean(Number(userBoard[y][x][0])),
                  })}
                  onClick={() => onSinkTry(x, y)}
                />
              ))}
            {!showUserBoard
              && Array.from({ length: size }).map((_, x) => (
                <div
                  key={`col-${x}`}
                  className={classnames(styles.column, {
                    [styles['enemy-1']]: board[y][x][0] === '1',
                    [styles['enemy-2']]: board[y][x][0] === '2',
                    [styles['enemy-3']]: board[y][x][0] === '3',
                    [styles['enemy-4']]: board[y][x][0] === '4',
                  })}
                />
              ))}
          </div>
        ))}
      </div>
      <button
        data-testid="toggle-board"
        onClick={() => setShowUserBoard(!showUserBoard)}
        className={styles.button}
      >
        {showUserBoard ? 'SHOW ENEMY BOARD' : 'SHOW USER BOARD'}
      </button>
      <button
        data-testid="save-game"
        onClick={() => handleSaveGame()}
        className={styles.button}
      >
        SAVE GAME
      </button>
    </div>
  );
}

Board.propTypes = {
  board: PropTypes.array.isRequired,
  userBoard: PropTypes.array.isRequired,
  onSinkTry: PropTypes.func.isRequired,
  onSaveGame: PropTypes.func.isRequired,
};
