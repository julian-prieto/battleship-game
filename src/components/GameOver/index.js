import PropTypes from "prop-types";

import gameover from "assets/gameover.png";

import styles from "./styles.module.scss";

export default function GameOver({ onGameReset: handleGameReset }) {
  return (
    <main className={styles.main}>
      <img src={gameover} width={450} alt="GameOver" />
      <button
        data-testid="reset"
        className={styles.button}
        onClick={() => handleGameReset()}
      >
        Play Again
      </button>
    </main>
  );
}

GameOver.propTypes = {
  onGameReset: PropTypes.func.isRequired,
};
