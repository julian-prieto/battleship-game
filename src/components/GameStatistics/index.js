import PropTypes from 'prop-types';

import styles from './styles.module.scss';

export default function GameStatistics({ game, onGameReset: handleGameReset }) {
  return (
    <div className={styles.main}>
      <span data-testid="level" className={styles.item}>
        Level:
        {' '}
        {game.difficulty.name}
      </span>
      <span data-testid="attempts" className={styles.item}>
        Attempts:
        {' '}
        {game.attempts}
        /
        {game.difficulty.maxAttempts || 'âˆž'}
      </span>
      <button
        data-testid="end-game"
        onClick={handleGameReset}
        className={styles.button}
      >
        End Game
      </button>
    </div>
  );
}

GameStatistics.propTypes = {
  game: PropTypes.object.isRequired,
  onGameReset: PropTypes.func.isRequired,
};
