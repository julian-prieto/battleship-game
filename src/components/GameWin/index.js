import PropTypes from 'prop-types';

import win from 'assets/win.png';

import styles from './styles.module.scss';

export default function GameWin({ onGameReset: handleGameReset }) {
  return (
    <main className={styles.main}>
      <span className={styles.title}>YOU WON!</span>
      <img src={win} width={450} alt="You Won!" />
      <button
        data-testid="reset"
        className={styles.button}
        onClick={() => handleGameReset()}
      >
        Play Again
      </button>
    </main>
  );
}

GameWin.propTypes = {
  onGameReset: PropTypes.func.isRequired,
};
