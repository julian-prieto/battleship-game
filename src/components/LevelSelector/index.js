import PropTypes from 'prop-types';
import { DIFFICULTY_LEVELS } from 'config';

import styles from './styles.module.scss';

export default function LevelSelector({ onSelectDifficulty }) {
  const handleSelectDifficulty = (level) => onSelectDifficulty(level);

  return (
    <main className={styles.component}>
      <span className={styles.title}>Please select difficulty</span>
      <div className={styles.buttons}>
        {DIFFICULTY_LEVELS.map((level) => (
          <button
            key={level.name}
            data-testid={`select-level-${level.name}`}
            className={styles.button}
            onClick={() => handleSelectDifficulty(level)}
          >
            {level.name}
          </button>
        ))}
      </div>
    </main>
  );
}

LevelSelector.propTypes = {
  onSelectDifficulty: PropTypes.func.isRequired,
};
