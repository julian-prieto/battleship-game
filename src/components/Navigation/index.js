import { Link } from 'react-router-dom';

import styles from './styles.module.scss';

export default function Navigation() {
  return (
    <nav>
      <ul data-testid="sections-list" className={styles.ul}>
        <li className={styles.li}>
          <Link to="/">Home</Link>
        </li>
        <li className={styles.li}>
          <Link to="/game">Play</Link>
        </li>
        <li className={styles.li}>
          <Link to="/scores">My Scores</Link>
        </li>
      </ul>
    </nav>
  );
}
