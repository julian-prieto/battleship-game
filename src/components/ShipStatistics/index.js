import PropTypes from "prop-types";
import classnames from "classnames";
import { DEFAULT_SHIPS } from "config";

import styles from "./styles.module.scss";

export default function ShipStatistics({ sunkShips }) {
  return (
    <div className={styles.main}>
      {DEFAULT_SHIPS.map((ship) => (
        <div
          data-testid={`ship-count-${ship.size}`}
          key={`ship-count-${ship.size}`}
          className={styles.item}
        >
          {Array.from({ length: ship.size }).map((_, index) => (
            <div
              key={`ship-draw-${ship.size}-${index}`}
              className={classnames(styles.ship, styles[`ship-${ship.size}`])}
            />
          ))}
          <span className={styles.value}>
            {sunkShips[ship.size] || 0}/{ship.amount}{" "}
          </span>
        </div>
      ))}
    </div>
  );
}

ShipStatistics.propTypes = {
  sunkShips: PropTypes.object.isRequired,
};
