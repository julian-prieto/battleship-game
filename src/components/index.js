export { default as Navigation } from './Navigation';
export { default as Board } from './Board';
export { default as LevelSelector } from './LevelSelector';
export { default as GameWin } from './GameWin';
export { default as GameOver } from './GameOver';
export { default as GameStatistics } from './GameStatistics';
export { default as ShipStatistics } from './ShipStatistics';
