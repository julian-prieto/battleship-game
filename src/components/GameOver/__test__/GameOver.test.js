import React from "react";
import ReactDOM from "react-dom";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { GameOver } from "components";

const handleGameReset = jest.fn();

it("Renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<GameOver onGameReset={handleGameReset} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("Renders Reset Button Correctly", () => {
  const { getByTestId } = render(<GameOver onGameReset={handleGameReset} />);
  expect(getByTestId("reset")).toHaveTextContent("Play Again");
});

it("Button onClick fires correctly", () => {
  const { getByTestId } = render(<GameOver onGameReset={handleGameReset} />);
  fireEvent.click(getByTestId("reset"));
  expect(handleGameReset).toHaveBeenCalledTimes(1);
});

it("Matches Snapshot", () => {
  const tree = renderer
    .create(<GameOver onGameReset={handleGameReset} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
