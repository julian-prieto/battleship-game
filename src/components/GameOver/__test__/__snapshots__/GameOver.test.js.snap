// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<main
  className="main"
>
  <img
    alt="GameOver"
    src="gameover.png"
    width={450}
  />
  <button
    className="button"
    data-testid="reset"
    onClick={[Function]}
  >
    Play Again
  </button>
</main>
`;
